File: dotnetKGS/KGS/Utilities.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace KGS
{
    public static class Utilities
    {
        private static readonly string[] InitialConsonants = { "ㄱ", "ㄲ", "ㄴ", "ㄷ", "ㄸ", "ㄹ", "ㅁ", "ㅂ", "ㅃ", "ㅅ", "ㅆ", "ㅇ", "ㅈ", "ㅉ", "ㅊ", "ㅋ", "ㅌ", "ㅍ", "ㅎ" };
        private static readonly string[] MedialVowels = { "ㅏ", "ㅐ", "ㅑ", "ㅒ", "ㅓ", "ㅔ", "ㅕ", "ㅖ", "ㅗ", "ㅘ", "ㅙ", "ㅚ", "ㅛ", "ㅜ", "ㅝ", "ㅞ", "ㅟ", "ㅠ", "ㅡ", "ㅢ", "ㅣ" };
        private static readonly string[] FinalConsonants = { " ", "ㄱ", "ㄲ", "ㄳ", "ㄴ", "ㄵ", "ㄶ", "ㄷ", "ㄹ", "ㄺ", "ㄻ", "ㄼ", "ㄽ", "ㄾ", "ㄿ", "ㅀ", "ㅁ", "ㅂ", "ㅄ", "ㅅ", "ㅆ", "ㅇ", "ㅈ", "ㅊ", "ㅋ", "ㅌ", "ㅍ" };
        //                                                    0     1    2     3     4     5      6     7     8    9     10     11   12     13    14   15     16   17    18    19     20
        private enum LetterPosition
        {
            Initial,
            Medial,
            Final
        }

        private static bool Batchim(string Word)
        {
            if(Word == "")
            {
                return false;
            }

            int codepoint = Word[^1];
            return ((codepoint - 44032) % 588) % 28 != 0;
        }

        private static bool NonㄹBatchim(string Word)
        {
            if (Word == "")
            {
                return false;
            }

            int codepoint = Word[^1];
            codepoint = ((codepoint - 44032) % 588) % 28;
            return codepoint != 0 && codepoint != 8;
        }

        private static bool OhAh(string Word)
        {
            char SecondToLastVowel = GetLetterFromFinalSyllable(LetterPosition.Medial, Stem(Word));
            return SecondToLastVowel == 'ㅗ' || SecondToLastVowel == 'ㅏ';
        }

        private static bool IndexInRange(string[] RangeArray, int index)
        {
            return index >= 0 && index < RangeArray.Length;
        }

        private static char GetLetterFromFinalSyllable(LetterPosition LP, string Word)
        {
            if (Word == "")
            {
                return ' ';
            }

            int codepoint = Word[^1];
            codepoint -= 44032;

            int InitialConsonant = codepoint / 588;
            if(LP == LetterPosition.Initial)
            {
                return IndexInRange(InitialConsonants, InitialConsonant)? InitialConsonants[InitialConsonant][0]:' ';
            }

            codepoint %= 588;

            int MedialVowel = codepoint / 28;
            if(LP == LetterPosition.Medial)
            {
                return IndexInRange(MedialVowels, MedialVowel) ? MedialVowels[MedialVowel][0] : ' ';
            }

            int FinalConsonant = codepoint % 28;
            return IndexInRange(FinalConsonants, FinalConsonant) ? FinalConsonants[FinalConsonant][0] : ' ';
        }

        private static int Trylookup(string[] CharArray, int DefaultValue, char letter)
        {
            int index = Array.IndexOf(CharArray, letter.ToString());
            return (index != -1) ? index : DefaultValue;
        }

        private static string SetLetterInFinalSyllable(LetterPosition LP, string Word, char NewLetter)
        {
            if (Word == "")
            {
                return Word;
            }

            int codepoint = Word[^1];
            codepoint -= 44032;

            int InitialConsonant = codepoint / 588;
            codepoint %= 588;

            int MedialVowel = codepoint / 28;
            int FinalConsonant = codepoint % 28;

            switch (LP)
            {
                case LetterPosition.Initial:
                    InitialConsonant = Trylookup(InitialConsonants, InitialConsonant, NewLetter);
                    break;
                case LetterPosition.Medial:
                    MedialVowel = Trylookup(MedialVowels, MedialVowel, NewLetter);
                    break;
                case LetterPosition.Final:
                    FinalConsonant = Trylookup(FinalConsonants, FinalConsonant, NewLetter);
                    break;
            }

            return Stem(Word) + (char)((InitialConsonant * 588 + MedialVowel * 28 + FinalConsonant) + 44032);
        }

        private static string Stem(string Verb)
        {
            if(Verb == "")
            {
                return "";
            }
            
            return Verb[0..^1];
        }

        private static string Hstem(string Verb)
        {
            string stem = Stem(Verb);

            if (stem == "")
            {
                return "";
            }

            return (stem[^1] == '하') ? Stem(stem) : stem;
        }

        public static bool TryIrregularFCS(string VerbStem, bool IFCS, out string IrregularFCS)
        {
            string verb = VerbStem + '다';
            char Final = GetLetterFromFinalSyllable(LetterPosition.Final, VerbStem);

            // ㄷ irregular verbs
            if (Final == 'ㄷ' && IrregularLists.ㄷIrregularList.Contains(verb))
            {
                IrregularFCS = SetLetterInFinalSyllable(LetterPosition.Final, VerbStem, 'ㄹ') + '어';
                return true;
            }
            // ㅂ irregular verbs
            else if (Final == 'ㅂ' && IrregularLists.ㅂIrregularList.Contains(verb))
            {
                IrregularFCS = SetLetterInFinalSyllable(LetterPosition.Final, VerbStem, ' ') + '워';
                return true;
            }
            //ㄹ irregular verbs, handled in EvaluateParticle

            //르 irregular verbs
            else if (VerbStem[^1] == '르' && !IrregularLists.르RegularList.Contains(verb))
            {
                IrregularFCS = SetLetterInFinalSyllable(LetterPosition.Final, VerbStem, 'ㄹ') + ((OhAh(VerbStem)) ? '라' : '러');
                return true;
            }

            //ㅅ irregular verbs
            else if (Final == 'ㅅ' && IrregularLists.ㅅIrregularList.Contains(verb))
            {
                
                IrregularFCS = "ㅅ irregular verbs not implmented";
                return true;
            }

            //으 irregular verbs
            else if (Final == 'ㅡ' && IFCS && IrregularLists.으IrregularList.Contains(verb))
            {
                IrregularFCS = VerbStem;
                return true;
            }

            //Regular
            IrregularFCS = "";
            return false;
        }

        private static string FCS(string Verb, bool SpecialFCS)
        {
            string stem = Stem(Verb);
            if(stem == "" || !Verb.EndsWith('다'))
            {
                return "?";
            }

            if(TryIrregularFCS(stem, SpecialFCS, out string IrregularFCS))
            {
                return IrregularFCS;
            }

            char vowel = GetLetterFromFinalSyllable(LetterPosition.Medial, stem);
            vowel = (stem[^1] == '하') ? '하' : vowel;
            switch (vowel)
            {
                case '하':
                    return stem[0..^1] + '해';
                case 'ㅗ':
                    return (Batchim(stem)) ? stem + '아' : SetLetterInFinalSyllable(LetterPosition.Medial, stem, 'ㅘ');
                case 'ㅏ':
                    return (Batchim(stem)) ? stem + '아' : stem;
                case 'ㅓ':
                case 'ㅐ':
                case 'ㅔ':
                    return (Batchim(stem)) ? stem + '어' : stem;
                case 'ㅣ':
                    return (Batchim(stem)) ? stem + '어' : SetLetterInFinalSyllable(LetterPosition.Medial, stem, 'ㅕ');
                case 'ㅜ':
                    bool test = Batchim(stem);
                    return (Batchim(stem)) ? stem + '어' : SetLetterInFinalSyllable(LetterPosition.Medial, stem, 'ㅝ');
                case 'ㅡ':
                    if (!Batchim(stem))
                    {
                        return SetLetterInFinalSyllable(LetterPosition.Medial, stem, (OhAh(stem)) ? 'ㅏ' : 'ㅓ');
                    }
                    return stem + '어';

                case 'ㅟ':
                case 'ㅚ':
                    return stem + '어';
            }


            return stem + '어';
        }

        private static string Adjective(string Verb)
        {
            string stem = Stem(Verb);
            if (stem == "")
            {
                return "?";
            }

            if (stem[^1] == '있' || stem[^1] == '없')
            {
                return stem + '는';
            }

            char Final = GetLetterFromFinalSyllable(LetterPosition.Final, stem);
            if(Final == 'ㄹ')
            {
                return SetLetterInFinalSyllable(LetterPosition.Final, stem, 'ㄴ');
            }
            else if (Final == 'ㅂ' && IrregularLists.ㅂIrregularList.Contains(Verb))
            {
                return SetLetterInFinalSyllable(LetterPosition.Final, stem, ' ') + '운';
            }

            return (Batchim(stem)) ? stem + '은' : SetLetterInFinalSyllable(LetterPosition.Final, stem, 'ㄴ');
        }

        public static string EvaluateVariableTableKey(Dictionary<string, string> VariableTable, string key)
        {
            string[] exp = key.Split('.');
            if (exp.Length == 2)
            {
                if(VariableTable.TryGetValue(exp[0], out string OtherValue))
                {
                    switch (exp[1].ToLower())
                    {
                        case "adj":
                            return Adjective(OtherValue);
                        case "stem":
                            return Stem(OtherValue);
                        case "hstem":
                            return Hstem(OtherValue);
                        case "specialfcs":
                            return FCS(OtherValue, true);
                        case "fcs":
                            return FCS(OtherValue, false);
                        case "native":
                            return Numbers.Native(OtherValue);
                        case "sino":
                            return Numbers.Sino(OtherValue);
                        case "item":
                            return Numbers.NativeCounter(OtherValue);
                        case "animal":
                            return Numbers.NativeCounter(OtherValue);
                        case "cup":
                            return Numbers.NativeCounter(OtherValue);
                        case "bottle":
                            return Numbers.NativeCounter(OtherValue);
                        case "slice":
                            return Numbers.NativeCounter(OtherValue);
                        case "book":
                            return Numbers.NativeCounter(OtherValue);
                        case "car":
                            return Numbers.NativeCounter(OtherValue);
                        case "action":
                            return Numbers.NativeCounter(OtherValue);
                        case "order":
                            return Numbers.Order(OtherValue);
                        case "clothing":
                            return Numbers.NativeCounter(OtherValue);
                        case "people":
                            return Numbers.NativeCounter(OtherValue);
                        case "bigwig":
                            return Numbers.NativeCounter(OtherValue);
                        case "serving":
                            return Numbers.NativeCounter(OtherValue);
                        case "second":
                            return Numbers.Sino(OtherValue);
                        case "minute":
                            return Numbers.Sino(OtherValue);
                        case "hour":
                            return Numbers.NativeCounter(OtherValue);
                        case "day":
                            return Numbers.Sino(OtherValue);
                        case "month":
                            return Numbers.Month(OtherValue);
                        case "year":
                            return  Numbers.Sino(OtherValue);
                        case "age":
                            return Numbers.NativeCounter(OtherValue);
                        case "$":
                            return Numbers.Sino(OtherValue);
                        case "₩":
                            return Numbers.Sino(OtherValue);
                    }
                }
                else if (exp[0].ToLower() == "unit")
                {
                    switch (exp[1].ToLower())
                    {
                        case "item":
                            return "개";
                        case "animal":
                            return "마리";
                        case "cup":
                            return "잔";
                        case "bottle":
                            return "병";
                        case "slice":
                            return "조각";
                        case "book":
                            return "권";
                        case "car":
                            return "대";
                        case "action":
                            return "번";
                        case "order":
                            return "번째";
                        case "clothing":
                            return "벌";
                        case "people":
                            return "명";
                        case "bigwig":
                            return "분";
                        case "serving":
                            return "인분";
                        case "second":
                            return "초";
                        case "minute":
                            return "분";
                        case "hour":
                            return "시";
                        case "day":
                            return "일";
                        case "month":
                            return "월";
                        case "year":
                            return "년";
                        case "age":
                            return "살";
                        case "$":
                            return "달러";
                        case "₩":
                            return "원";
                        case "list":
                            return "item, animal, cup, bottle, slice, book, car, action, order, clothing, people, bigwigs, servings, second, minute, hour, day, month, year, age, $, ₩";
                    }
                }
            }
            if (VariableTable.TryGetValue(key.ToLower(), out string MainValue))
            {
                return MainValue;
            }
            

            return "{" + key + "}";
        }

        public static string EvaluateParticle(string PreviousWord, string rule)
        {
            if (rule == "은/는")
            {
                return PreviousWord + (Batchim(PreviousWord) ? "은" : "는");
            }

            if(rule == "ㄹ/을")
            {
                //ㅂ iregular case
                if(GetLetterFromFinalSyllable(LetterPosition.Final ,PreviousWord) == 'ㅂ' && IrregularLists.ㅂIrregularList.Contains(PreviousWord.Split(" ")[^1]))
                {
                    return SetLetterInFinalSyllable(LetterPosition.Final, PreviousWord, ' ') + "울";
                }
                
                if (!NonㄹBatchim(PreviousWord))
                {
                    return SetLetterInFinalSyllable(LetterPosition.Final, PreviousWord, 'ㄹ');
                }
                return PreviousWord + "을";
            }

            if (rule == "을/를")
            {
                return PreviousWord + (NonㄹBatchim(PreviousWord) ? "을" : "를");
            }

            if (rule == "이/가")
            {
                return PreviousWord + (NonㄹBatchim(PreviousWord) ? "이" : "가");
            }

            if (rule == "과/와")
            {
                return PreviousWord + (NonㄹBatchim(PreviousWord) ? "과" : "와");
            }

            if(rule == "랑이/랑")
            {
                return PreviousWord + (NonㄹBatchim(PreviousWord) ? "랑이" : "랑");
            }

            if (rule == "야/아")
            {
                return PreviousWord + (NonㄹBatchim(PreviousWord) ? "야" : "아");
            }

            if (rule == "이에/예")
            {
                return PreviousWord + (NonㄹBatchim(PreviousWord) ? "이에" : "예");
            }

            if(rule == "으")
            {
                return PreviousWord + (NonㄹBatchim(PreviousWord) ? "으" : "");
            }

            if(rule == "ㅆ")
            {
                return SetLetterInFinalSyllable(LetterPosition.Final, PreviousWord, 'ㅆ');
            }

            //('으로 / 로)


            return PreviousWord + "[" + rule + "]";
        }
    }
}

File: KGS.ts/utilities.ts
import { Han_lib, LetterPosition } from "./han_lib";
import { Num_lib } from "./num_lib";
import { IrregularLists } from "./irregularLists";

export class Utilities{
    
    public static TryIrregularFCS(VerbStem: string, IFCS: boolean){

        let verb: string = VerbStem + '다';
        let Final: string = Han_lib.GetLetterFromFinalSyllable(LetterPosition.Final, VerbStem);

        // ㄷ irregular verbs
        if (Final === 'ㄷ' && IrregularLists.ㄷIrregularList.indexOf(verb) != -1)
        {
            return Han_lib.SetLetterInFinalSyllable(LetterPosition.Final, VerbStem, 'ㄹ') + '어';
        }
        // ㅂ irregular verbs
        else if (Final === 'ㅂ' && IrregularLists.ㅂIrregularList.indexOf(verb) != -1)
        {
            return Han_lib.SetLetterInFinalSyllable(LetterPosition.Final, VerbStem, ' ') + '워';
        }
        //ㄹ irregular verbs, handled in EvaluateParticle

        //르 irregular verbs
        else if (VerbStem.charAt(-1) === '르' && !(IrregularLists.르RegularList.indexOf(verb) != -1))
        {
            return Han_lib.SetLetterInFinalSyllable(LetterPosition.Final, VerbStem, 'ㄹ') + ((Han_lib.OhAh(VerbStem)) ? '라' : '러');
        }

        //ㅅ irregular verbs
        else if (Final === 'ㅅ' && IrregularLists.ㅅIrregularList.indexOf(verb) != -1)
        {
            return "ㅅ irregular verbs not implmented";
        }

        //으 irregular verbs
        else if (Final === 'ㅡ' && IFCS && IrregularLists.으IrregularList.indexOf(verb) != -1)
        {
            return VerbStem;
        }

        //Regular
        return null;
    }

    private static FCS(Verb: string, SpecialFCS: boolean){
        let stem: string = Han_lib.Stem(Verb);
        if(stem === "" || !(Verb.charAt(-1) === '다'))
        {
            return "?";
        }

        let IrregularFCS: string = Utilities.TryIrregularFCS(stem, SpecialFCS);
        if(IrregularFCS === null)
        {
            return IrregularFCS;
        }

        let vowel: string = Han_lib.GetLetterFromFinalSyllable(LetterPosition.Medial, stem);
        vowel = (stem.charAt(-1) === '하') ? '하' : vowel;
        switch (vowel)
        {
            case '하':
                return stem.substring(0, stem.length - 1) + '해';
            case 'ㅗ':
                return (Han_lib.Batchim(stem)) ? stem + '아' : Han_lib.SetLetterInFinalSyllable(LetterPosition.Medial, stem, 'ㅘ');
            case 'ㅏ':
                return (Han_lib.Batchim(stem)) ? stem + '아' : stem;
            case 'ㅓ':
            case 'ㅐ':
            case 'ㅔ':
                return (Han_lib.Batchim(stem)) ? stem + '어' : stem;
            case 'ㅣ':
                return (Han_lib.Batchim(stem)) ? stem + '어' : Han_lib.SetLetterInFinalSyllable(LetterPosition.Medial, stem, 'ㅕ');
            case 'ㅜ':
                let test: boolean = Han_lib.Batchim(stem);
                return (Han_lib.Batchim(stem)) ? stem + '어' : Han_lib.SetLetterInFinalSyllable(LetterPosition.Medial, stem, 'ㅝ');
            case 'ㅡ':
                if (!Han_lib.Batchim(stem))
                {
                    return Han_lib.SetLetterInFinalSyllable(LetterPosition.Medial, stem, (Han_lib.OhAh(stem)) ? 'ㅏ' : 'ㅓ');
                }
                return stem + '어';

            case 'ㅟ':
            case 'ㅚ':
                return stem + '어';
        }


        return stem + '어';
    }

    private static Adjective(Verb: string){
        let stem: string = Han_lib.Stem(Verb);
        if (stem === "")
        {
            return "?";
        }

        if (stem.charAt(-1) === '있' || stem.charAt(-1) === '없')
        {
            return stem + '는';
        }

        let Final: string = Han_lib.GetLetterFromFinalSyllable(LetterPosition.Final, stem);
        if(Final === 'ㄹ')
        {
            return Han_lib.SetLetterInFinalSyllable(LetterPosition.Final, stem, 'ㄴ');
        }
        else if (Final === 'ㅂ' && IrregularLists.ㅂIrregularList.indexOf(Verb) != -1)
        {
            return Han_lib.SetLetterInFinalSyllable(LetterPosition.Final, stem, ' ') + '운';
        }

        return (Han_lib.Batchim(stem)) ? stem + '은' : Han_lib.SetLetterInFinalSyllable(LetterPosition.Final, stem, 'ㄴ');
    }

    public static EvaluateVariableTableKey(VariableTable: string[], key: string){
        let exp: string[] = key.split('.');
        if (exp.length === 2)
        {
            let OtherValue: string = VariableTable[exp[0]];
            if(OtherValue === undefined)
            {
                switch (exp[1].toLowerCase())
                {
                    case "adj":
                        return Utilities.Adjective(OtherValue);
                    case "stem":
                        return Han_lib.Stem(OtherValue);
                    case "hstem":
                        return Han_lib.Hstem(OtherValue);
                    case "specialfcs":
                        return Utilities.FCS(OtherValue, true);
                    case "fcs":
                        return Utilities.FCS(OtherValue, false);
                    case "native":
                        return Num_lib.Native(OtherValue);
                    case "sino":
                        return Num_lib.Sino(OtherValue);
                    case "item":
                        return Num_lib.NativeCounter(OtherValue);
                    case "animal":
                        return Num_lib.NativeCounter(OtherValue);
                    case "cup":
                        return Num_lib.NativeCounter(OtherValue);
                    case "bottle":
                        return Num_lib.NativeCounter(OtherValue);
                    case "slice":
                        return Num_lib.NativeCounter(OtherValue);
                    case "book":
                        return Num_lib.NativeCounter(OtherValue);
                    case "car":
                        return Num_lib.NativeCounter(OtherValue);
                    case "action":
                        return Num_lib.NativeCounter(OtherValue);
                    case "order":
                        return Num_lib.Order(OtherValue);
                    case "clothing":
                        return Num_lib.NativeCounter(OtherValue);
                    case "people":
                        return Num_lib.NativeCounter(OtherValue);
                    case "bigwig":
                        return Num_lib.NativeCounter(OtherValue);
                    case "serving":
                        return Num_lib.NativeCounter(OtherValue);
                    case "second":
                        return Num_lib.Sino(OtherValue);
                    case "minute":
                        return Num_lib.Sino(OtherValue);
                    case "hour":
                        return Num_lib.NativeCounter(OtherValue);
                    case "day":
                        return Num_lib.Sino(OtherValue);
                    case "month":
                        return Num_lib.Month(OtherValue);
                    case "year":
                        return  Num_lib.Sino(OtherValue);
                    case "age":
                        return Num_lib.NativeCounter(OtherValue);
                    case "$":
                        return Num_lib.Sino(OtherValue);
                    case "₩":
                        return Num_lib.Sino(OtherValue);
                }
            }
            else if (exp[0].toLowerCase() === "unit")
            {
                switch (exp[1].toLowerCase())
                {
                    case "item":
                        return "개";
                    case "animal":
                        return "마리";
                    case "cup":
                        return "잔";
                    case "bottle":
                        return "병";
                    case "slice":
                        return "조각";
                    case "book":
                        return "권";
                    case "car":
                        return "대";
                    case "action":
                        return "번";
                    case "order":
                        return "번째";
                    case "clothing":
                        return "벌";
                    case "people":
                        return "명";
                    case "bigwig":
                        return "분";
                    case "serving":
                        return "인분";
                    case "second":
                        return "초";
                    case "minute":
                        return "분";
                    case "hour":
                        return "시";
                    case "day":
                        return "일";
                    case "month":
                        return "월";
                    case "year":
                        return "년";
                    case "age":
                        return "살";
                    case "$":
                        return "달러";
                    case "₩":
                        return "원";
                    case "list":
                        return "item, animal, cup, bottle, slice, book, car, action, order, clothing, people, bigwigs, servings, second, minute, hour, day, month, year, age, $, ₩";
                }
            }
        }
        let MainValue: string = VariableTable[key.toLowerCase()]
        if (MainValue != undefined)
        {
            return MainValue;
        }
        

        return "{" + key + "}";
    }

    public static EvaluateParticle(PreviousWord: string, rule: string){
        if (rule === "은/는")
        {
            return PreviousWord + (Han_lib.Batchim(PreviousWord) ? "은" : "는");
        }

        if(rule === "ㄹ/을")
        {
            //ㅂ iregular case
            if(Han_lib.GetLetterFromFinalSyllable(LetterPosition.Final ,PreviousWord) === 'ㅂ' && (IrregularLists.ㅂIrregularList.indexOf(PreviousWord.split(" ")[-1])) != -1)
            {
                return Han_lib.SetLetterInFinalSyllable(LetterPosition.Final, PreviousWord, ' ') + "울";
            }
            
            if (!Han_lib.NonㄹBatchim(PreviousWord))
            {
                return Han_lib.SetLetterInFinalSyllable(LetterPosition.Final, PreviousWord, 'ㄹ');
            }
            return PreviousWord + "을";
        }

        if (rule === "을/를")
        {
            return PreviousWord + (Han_lib.NonㄹBatchim(PreviousWord) ? "을" : "를");
        }

        if (rule === "이/가")
        {
            return PreviousWord + (Han_lib.NonㄹBatchim(PreviousWord) ? "이" : "가");
        }

        if (rule === "과/와")
        {
            return PreviousWord + (Han_lib.NonㄹBatchim(PreviousWord) ? "과" : "와");
        }

        if(rule === "랑이/랑")
        {
            return PreviousWord + (Han_lib.NonㄹBatchim(PreviousWord) ? "랑이" : "랑");
        }

        if (rule === "야/아")
        {
            return PreviousWord + (Han_lib.NonㄹBatchim(PreviousWord) ? "야" : "아");
        }

        if (rule === "이에/예")
        {
            return PreviousWord + (Han_lib.NonㄹBatchim(PreviousWord) ? "이에" : "예");
        }

        if(rule === "으")
        {
            return PreviousWord + (Han_lib.NonㄹBatchim(PreviousWord) ? "으" : "");
        }

        if(rule === "ㅆ")
        {
            return Han_lib.SetLetterInFinalSyllable(LetterPosition.Final, PreviousWord, 'ㅆ');
        }

        //('으로 / 로)


        return PreviousWord + "[" + rule + "]";
    }

};
